### PR TITLE
Fix joins to many channels

### PR DESCRIPTION
**Pull request by cboylan**
If you have enough channels that the 512 byte message limit on the JOIN
message is hit then limnoria was losing the channel that put it over the
limit and not including it in the next JOIN message. This resulted in
losing one channel for every JOIN message that pushed us over 512 bytes.

We fix this by generating the JOIN message immediately after resetting
the channels list to ensure we include the channel that pushed us over
the limit. Then the next time through our JOIN msg construction we'll
add subsequent channels without forgetting the one that pushed us over.

### CODE REVIEW

**Comment by ProgVal** [comment]
Oops. could you add a regression test in `tests/test_conf.py`?

**Comment by fungi** [comment]
Thanks for pushing this. I've applied it manually in our deployment and it's corrected the behavior. We configure a bot to join 88 channels presently, and this was getting split into multiple messages causing us to skip joining a random 2 channels at the boundaries between the messages. Restarting with this applied we joined all 88 correctly.

File: test/test_conf.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            name = "#verylongchannelname" + str(x)
+            channels.add(name)
+            input_list.append(name)
+        conf.supybot.networks.test.channels.setValue(channels)
+        msgs = conf.supybot.networks.test.channels.joins()
+        # Double check we split the messages
+        self.assertEqual(len(msgs), 2)
+        # Ensure all channel names are present
+        chan_list = (msgs[0].args[0] + ',' + msgs[1].args[0]).split(',')
+        self.assertEqual(sorted(input_list), sorted(chan_list))
```
Comment (ProgVal) [inline test/test_conf.py:58]: `         self.assertCountEqual(input_list, chan_list)

**Comment by ProgVal** [comment]
what did you change?

**Comment by cboylan** [comment]
I added the requested `self.assertCountEqual(input_list, chan_list)`.

**Comment by ProgVal** [comment]
Oh I missed it. I meant it as a replacement for the comparison between sorted()

**Comment by cboylan** [comment]
I can make that change, though it seems that comparing the actual channel names given as input against the channels listed in the output is also a good check? The count would catch this particular issue, but it is possible some other transformation could happen that you would want to guard against?

**Comment by ProgVal** [comment]
`assertCountEqual` does compare the values, not just the lengths: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual

**Comment by cboylan** [comment]
> `assertCountEqual` does compare the values, not just the lengths: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual

Aha. Today I learned. Commit has been updated to do only assertCountEqual().

**Comment by ProgVal** [comment]
Thanks! :)

**Pull request merged**